service.go

```diff
@@ -0,0 +1,337 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apiError "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Struct to build service watcher which looks after target services.
+type ServiceWatcher struct {
+	ctx       Context
+	svcFilter labels.Selector
+	informer  cache.SharedInformer
+	// Name space where to install service watcher,
+	namespace string
+}
+
+// Build new service watcher
+func NewServiceWatcher(ctx Context, ns *string) ServiceWatcher {
+	svc := &ServiceWatcher{
+		ctx: ctx,
+	}
+	svc.svcFilter = svc.createServiceFilter()
+	svc.informer = svc.createSharedInformer()
+	svc.namespace = *ns
+	return *svc
+}
+
+func (svc *ServiceWatcher) checkIfNameSpaceExists() {
+
+	// Check if the namespace already exists
+	client := svc.ctx.client
+	log := svc.ctx.log
+	namespace := svc.namespace
+	_, err := client.CoreV1().Namespaces().Get(svc.ctx.ctx, namespace, metav1.GetOptions{})
+	if apiError.IsAlreadyExists(err) {
+		log.Debugf("Namespace '%s' already exists", namespace)
+		return
+	} else {
+		// Create the namespace
+		newNamespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespace,
+			},
+		}
+		_, err := client.CoreV1().Namespaces().Create(svc.ctx.ctx, newNamespace, metav1.CreateOptions{})
+		if err != nil {
+			log.Errorf("Issue creating namesapce : %v", err.Error())
+		}
+
+		log.Infof("Namespace '%s' created", namespace)
+	}
+}
```

Comment:
This is good, but what if the error is actually not nil here? We'll still create a namespace. However, if our error is related to rbac, or a transient network failure, socket's been dropped, etc (you get it), then we'll make it harder on ourselves to actually isolate the root issue.

```suggestion
    if err != nil {
        if apiError.IsAlreadyExists(err) {
		      log.Debugf("Skipped creating namespace '%s'; already exists", namespace)
		      return
		}
		log.Errorf("Failed to get namespace '%s': %v", err)
		return
	} 
     // Create the namespace
	newNamespace := &corev1.Namespace{
			ObjectMeta: metav1.ObjectMeta{
				Name: namespace,
			},
		}
		_, err := client.CoreV1().Namespaces().Create(svc.ctx.ctx, newNamespace, metav1.CreateOptions{})
		if err != nil {
			log.Errorf("Issue creating namespace : %v", err.Error())
		}

		log.Infof("Namespace '%s' created", namespace)
	}
}
```

It's also generally better to return early instead of nesting in ifs